tools/wptrunner/wptrunner/testloader.py

```diff
@@ -270,32 +271,42 @@ class TestFilter:
     """Callable that restricts the set of tests in a given manifest according
     to initial criteria"""
     def __init__(self, test_manifests, include=None, exclude=None, manifest_path=None, explicit=False):
-        if manifest_path is None or include or explicit:
-            self.manifest = manifestinclude.IncludeManifest.create()
-            self.manifest.set_defaults()
-        else:
-            self.manifest = manifestinclude.get_manifest(manifest_path)
-
-        if include or explicit:
-            self.manifest.set("skip", "true")
-
+        self.include_urls = set()
         if include:
-            for item in include:
-                self.manifest.add_include(test_manifests, item)
+            for include_pattern in include:
+                self.include_urls.update(manifestinclude.resolve_pattern(test_manifests, include_pattern))
+        else:
+            if manifest_path:
+                manifest = manifestinclude.get_manifest(manifest_path)
+            else:
+                manifest = manifestinclude.IncludeManifest.create()
+            manifest.set("skip", str(explicit).lower())
+            for test_manifest in test_manifests:
+                for _, _, tests in test_manifest:
+                    for test in tests:
+                        if not hasattr(test, "url"):
+                            continue
+                        if manifest.include(test):
+                            self.include_urls.add(test.id)
+
+        for exclude_pattern in (exclude or []):
+            self.include_urls -= set(manifestinclude.resolve_pattern(test_manifests, exclude_pattern))
 
-        if exclude:
-            for item in exclude:
-                self.manifest.add_exclude(test_manifests, item)
+    def __call__(self, manifest):
+        tests_by_url = group_tests_by_url(manifest)
+        tests_by_path = defaultdict(set)
+        for include_url in self.include_urls:
+            test = tests_by_url.get(include_url)
+            if test:
+                tests_by_path[test.path].add(test)
+        for test_path, tests in tests_by_path.items():
+            test_type = next(iter(tests)).item_type
+            yield test_type, test_path, tests
 
-    def __call__(self, manifest_iter):
-        for test_type, test_path, tests in manifest_iter:
-            include_tests = set()
-            for test in tests:
-                if self.manifest.include(test):
-                    include_tests.add(test)
 
-            if include_tests:
-                yield test_type, test_path, include_tests
+@functools.lru_cache(maxsize=16)
+def group_tests_by_url(manifest):
```

Comment:
Not sure why this is called *group*_tests_by_url? Also this is test_by_url instead of tests_by_url, right?

Reply:
I agree `group*` is a misnomer, which suggests some collection as a value. Changed to `map_tests_by_url` (it's not `map_test_by_url` because the map holds multiple tests, even though the value is a single test).